rpmio/rpmglob.c

```diff
@@ -665,6 +665,20 @@ void rpmUnescape(char *s)
     *p = '\0';
 }
 
+void rpmUnescapeQuotes(char *s)
+{
+    char *p, *q;
+    int esc = 0;
+    p = q = s;
+    while (*q != '\0') {
+	*p = *q++;
+	esc = (*p == '\\') && !esc && *q == '"';
```

Comment:
```suggestion
	esc = (*p == '\\') && !esc;
```

Reply:
Can you elaborate? Note that this is meant to specifically unescape quotes only. Also, it's mostly just a modified copy&paste of the preceding `rpmUnescape()` function at the moment (refactoring will come later).

Reply:
One of the goals of a "draft" PR is to "get early feedback", so I guess that's to be expected, OTOH this really is just a WIP.

Reply:
> Can you elaborate? Note that this is meant to specifically unescape quotes only. Also, it's mostly just a modified copy&paste of the preceding `rpmUnescape()` function at the moment (refactoring will come later).

Ah okay, that makes sense.  Other backslashes need to be preserved for `glob()`.  Is `"` the only character that needs to be unescaped, though?

Reply:
OK, let me explain the logic here:

The function `parseForSimple()` accepts quotes as a way to escape spaces (normally, multiple filenames are accepted on a single line in the `%files` section, each separated by whitespace). The tokenization into filenames is done in a loop by the `strtokWithQuotes()` function here, which is *where the quote semantics apply*. Once that's done, any quotes are supposed to have been consumed and so they no longer have any meaning and thus should be stripped; for the outermost quotes, this is already done by `strtokWithQuotes()` because it returns a quoted token without the opening and closing quotes. IOW, the *scope* of the quotes here is filename tokenization (the `strtokWithQuotes()` call). Any such quotes should be discarded in the resulting token (a filename) before we pass it further.

Now, backslashes used to denote literal quotes have the same scope as normal quotes; they are only meaningful for filename tokenization, not beyond that. So the same applies as above; discard such backslashes ("unescape") as soon as we're done tokenizing (we want just the quotes in the resulting filename(s), not the backslashes preceding them).

So we only want to unescape quotes here, not anything else, because we're already processing a single filename which is enclosed in quotes; that's the (non-obvious) `(s != t)` check. If there are other escaped chars in the filename, that's OK and we ignore them, because we do want those to be literal, as that's the semantics of quoted strings.

For example:
```
%files
"/foo\"bar"
"/foo\[bar"
```
These should be processed and eventually stat-ed as follows: `/foo"bar` and `/foo\[bar`.

I hope that makes sense :)

Reply:
How do you plan to handle these cases?

```
%files
'/foo'\''bar' # single quotes
'/foo'bar # partially quoted word
'/foo'* # mixture of quotes and globs
'/a \\ "\ ' # single quotes treat everything literally
# missing end quotes
'abc
"def
# backslash at end of line
a\
```

Some of those are just syntax errors, but at least RPM needs to report the error properly.

Reply:
Single quotes are currently not supported and I don't plan to handle those in any special way, really. That would just make things a lot more complicated that they currently are.

Partially quoted words and trailing backslashes, those are indeed syntax errors, I may take a look at those after I'm done with everything else, but it's slightly outside of scope here.

Reply:
> Single quotes are currently not supported and I don't plan to handle those in any special way, really. That would just make things a lot more complicated than they currently are.

In the interest of avoiding confusion, I suggest reporting an error if you encounter them.  Rejecting unescaped backslashes would also be a good idea; they are considered bad practice in shell scripts.

> Partially quoted words and trailing backslashes, those are indeed syntax errors, I may take a look at those after I'm done with everything else, but it's slightly out of scope here.

Why are partially quoted words syntax errors?  That means that quotes and globs cannot be mixed, which is annoying.

Reply:
> In the interest of avoiding confusion, I suggest reporting an error if you encounter them. Rejecting unescaped backslashes would also be a good idea; they are considered bad practice in shell scripts.

Fair enough.

> Why are partially quoted words syntax errors? That means that quotes and globs cannot be mixed, which is annoying.

Well, RPM currently only supports the basic form of quotation, which is `"/foo"`, i.e. an opening and closing quote at the very beginning and end of the line, respectively. What do you mean by "quotes and globs cannot be mixed"? Can you provide an example?

Reply:
Oh, I think I understand what you meant now:
`"/foo bar"[baz]`
Yes, that's not supported, although it *might* be easy enough to be added, but I don't know for certain without thinking about it more deeply.

The point is, in such a case, you're supposed to escape whatever you're escaping in that quoted part, with backslashes.

Reply:
> What do you mean by "quotes and globs cannot be mixed"? Can you provide an example?

`"/foo bar"/*`

Reply:
That's equivalent to the example I gave above.